homeassistant/components/recorder/purge.py

```diff
@@ -44,6 +44,23 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
 
             _LOGGER.debug("Purging states and events before %s", batch_purge_before)
 
+            # Update old_state_id to NULL before deleting to ensure
+            # the delete does not fail due to a foreign key constraint
+            # since some databases (MSSQL) cannot do the ON DELETE CASCADE
+            # for us.
+            disconnected_rows = (
+                session.query(States)
+                .filter(
+                    States.old_state_id.in_(
+                        session.query(States.state_id)
+                        .filter(States.last_updated < batch_purge_before)
+                        .subquery()
+                    )
```

Comment:
I would probably not do it that way. Say you would delete 10000 old states per day, combined with 10 days retention,  the result of the `in_` query would be
```
10000 states / day * 10 days * 10000 states to delete
```
That could be quite a loot to do in **one** query. Doing it in subsets would at least limit the amount of comparisons per query to a set maximum. From my recorder PR: https://github.com/home-assistant/core/blob/a665f4777c136e00f45635c29186325a3fd0872f/homeassistant/components/recorder/purge.py#L278-L285

Reply:
Isn't it limited to the size of the batch? Its the exact same query we are about to delete below?

```
            deleted_rows = (
                session.query(States)
                .filter(States.last_updated < batch_purge_before)
                .delete(synchronize_session=False)
            )
```

Reply:
I setup https://github.com/bdraco/scaletest to flood the state machine and then did the purge.

It handled 17k states without issue. The `delete` query took ~5s the new `update` query took ~3s.  The timings are based on the log messages but that seems good enough for testing this.

```
2021-02-17 06:56:18 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging states and events before target 2021-02-17 06:56:18.909061+00:00
2021-02-17 06:56:18 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging states and events before 2021-02-17 05:44:12+00:00
2021-02-17 06:56:18 DEBUG (Recorder) [homeassistant.components.recorder.purge] Updated 1060 states to remove old_state_id
2021-02-17 06:56:19 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 1257 states
2021-02-17 06:56:20 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 1677 events
2021-02-17 06:56:20 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging hasn't fully completed yet
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging states and events before target 2021-02-17 06:56:21.798652+00:00
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging states and events before 2021-02-17 06:44:14+00:00
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Updated 982 states to remove old_state_id
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 994 states
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 1006 events
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging hasn't fully completed yet
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging states and events before target 2021-02-17 06:56:21.974373+00:00
2021-02-17 06:56:21 DEBUG (Recorder) [homeassistant.components.recorder.purge] Purging states and events before 2021-02-17 06:56:21.974373+00:00
2021-02-17 06:56:24 DEBUG (Recorder) [homeassistant.components.recorder.purge] Updated 17790 states to remove old_state_id
2021-02-17 06:56:29 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 17922 states
2021-02-17 06:56:41 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 17930 events
2021-02-17 06:56:41 DEBUG (Recorder) [homeassistant.components.recorder.purge] Deleted 1 recorder_runs
2021-02-17 06:56:41 DEBUG (Recorder) [homeassistant.components.recorder.purge] Optimizing SQL DB to free space
```

Reply:
@bdraco `17k` events/states is not a big number :) My current database has `850k` events/states. 

In such cases, we should remove and update in batches, likely of 1k or 10k max. Long executions will lock the rest of HASS when trying to push data, as usually very large `DELETE` can lock the whole table.

Reply:
> Isn't it limited to the size of the batch since its the exact same query (less the delete) we are about to delete below (existing code)? The number of results should always be less than or exactly the same as the delete query which is more expensive and seems to be ok with the current design.

I was more worried about the `in_` clause here, since we need to check **all** remaining state if any `old_state_id` matches one that is about to be deleted. The update itself isn't a problem.

> There is an index on `old_state_id` so it should quite fast.

> It handled 17k states without issue. The delete query took ~5s the new update query took ~3s. The timings are based on the log messages but that seems good enough for testing this.

As I said, we would check the 17k states that are to be deleted against the remaining 100k+ states in the database. Might not be an issue, but I just wanted to suggest it for lower powered devices (Raspberry PI, ...). Cloud be fine with the index though.

Reply:
Usually, when you want to update/delete a lot of records, doing `DELETE` makes everything to block. The best is to chunk in a smaller batches to avoid contention.

I did not test that, but I would propose something like that, where-as we process data in a small chunks:

```python
while True:
  ids = (
    session.query(States)
      .filter(States.last_updated < batch_purge_before)
      .limit(10000)
      .query(States.id)
      .all()
  )

  if not ids:
    break

  disconnected_rows = (
    session.query(States)
      .filter(States.old_state_id.in_(ids))
      .update({"old_state_id": None}, synchronize_session=False)
  )
  _LOGGER.debug("Updated %s states to remove old_state_id", disconnected_rows)
  
  deleted_rows = (
      session.query(States)
      .filter(States.id.in_(ids))
      .delete(synchronize_session=False)
  )
  _LOGGER.debug("Deleted %s states", deleted_rows)
```

What it does:
- fetch a 10k of matching IDs (order is not important)
- update all newer states referencing these IDs to `old_state_id=NULL`
- delete them (`NO ACTION`) will be safe
- iterate until the list is empty

Reply:
@ayufan We already limit the amount of actions, just not by count. The purge happens in hour increments.

Reply:
> The purge happens in hour increments

@cdce8p 

For ex. I keep history for 3 days. It means that each 1h results in removal of 12k events/states. I would not be surprised that there will be a people that would be removing more per hour. What I'm trying to say, someone could have a backlog that will create a random slownesses on the whole system in regular intervals.

Reply:
I don't know about you, but I would image that someone who has that many states per hour doesn't run Home Assistant on a Raspberry PI but instead uses something more performant. Since the automatic purge happens at night, I wouldn't be too concerned about it.

Reply:
@cdce8p 

OK

I personally run on Turris Omnia. It is slower than Raspberry PI 4 :)

Reply:
That's impressive that the system can handle that many states changes per day.

While changing the way the purge is batched from by hour to by chunk size will like improve the responsiveness on slower systems, it's outside the scope of this PR.  

If that's something you want to take on, it seems like a good case for a follow up PR

Reply:
At that number of state changes, it might even make sense to use `multiprocessing` to run the recorder in another process and setup a pipe to push events to recorder to it.  This would side step the GIL and ensure there was minimal impact on the main process.

Reply:
The above solution would also avoid the need for transaction. That really are needed in the old solution. The set of selected states could change between the removal of old_state_id and the actual deletion.

Reply:
Thats definitely a problem, probably the whole purge process needs to get refactored to handle that case.

Reply:
I'll rework this next week when I have a bit more time as we will have to start over with the testing.

Reply:
I switched this to do batches of 5000.  Unfortunately, its basically a complete re-write of the purge logic.